### ISSUE TITLE
[Feature Request] AsusRouter scan method

### ISSUE THREAD

**Issue by Th3D4rk0n3** [leiweibau/Pi.Alert#177]
Hello!
Is it possible to add Asus router integration to Pi.Alert that would read the DHCP lease similar to Mikrotik and Unifi?

**Comment by leiweibau** [comment]
Since I don't have an ASUS router myself, I have to rely on pull requests for integration.

**Comment by tteck** [comment]
![Screenshot 2023-10-01 10 17 52 AM](https://github.com/leiweibau/Pi.Alert/assets/21342670/f155a3a6-68de-4819-bd55-1ccd47961ea1)

**Comment by jeeftor** [comment]
I have an asus but dont know if I have the time... If running merlin:

just have to grab: `/var/lib/misc/dnsmasq.leases`

**Comment by jeeftor** [comment]
Looks like there is a home assistant library I may be able to use to PR this in... 

Locally I got the lib to log into the router and pull down the network table...

**Comment by mdima** [comment]
> I have an asus but dont know if I have the time... If running merlin:
> 
> just have to grab: `/var/lib/misc/dnsmasq.leases`

Well, if you use a different DNS this won't work. 

Grabbing the device list from an http request (like the Home Assistant integration) will succeed instead!

If you want to test this feature I will be happy to help. 

Thanks,
Michele

**Comment by jeeftor** [comment]
I actually looked at the home assistant library they are using which does basically a login and caches the cookie and then hits a cgi script to pull down the Network list. 

I found the data and then didn't go any further than that...


```python
import asyncio
import json

import aiohttp

from asusrouter import AsusRouter, Endpoint


async def main():

    async with aiohttp.ClientSession() as session:
        router = AsusRouter(
            hostname="192.168.1.1",      # required - both IP and URL supported
            username="admin",            # required
            password="MY_ROUTER_PASSWORD",      # required
            port=80,                   # optional - default port based on use_ssl
            use_ssl=False,               # optional
            cache_time=5 ,                # optional
            session=session
        )

        connection = await router.async_connect()
        id = await router.async_get_identity()

        result = await router.async_api_hook("get_clientlist()")
        fields = ['name','nickName','ip','mac','vendor','ipMethod']

        extracted_data = [{key: client[key] for key in fields if key in client} for client in
                          result['get_clientlist'].values()]

        print(json.dumps(extracted_data))


# Run the main function in an async context
asyncio.run(main())
```

**Comment by leiweibau** [comment]
Thank you for your efforts so far. That looks good so far. This could certainly be helpful for a future pull request. Unfortunately, I still can't help here myself, as I still don't have an Asus router and don't want to roll out code without testing.

**Comment by mdima** [comment]
That is a great result, just I do not have any experience in Python, I can try to integrate it but will take ages... :S

**Comment by jeeftor** [comment]
Its a time issue for me ... I have work plus a bunch of other home assistant stuff I'm working on...

Really we just need to take the output of my function there and put it into the right data structure for SQL stuff... if somebody pings me in a few weeks maybe I'll bang it out

**Comment by jeeftor** [comment]
I know this is a fork - but the codebase is ... well a MEGA file: https://github.com/leiweibau/Pi.Alert/blob/main/back/pialert.py

I'm worried if I was to make a PR I'd try to refactor the shit out of it and that would take even more of my time :)

**Comment by leiweibau** [comment]
> I know this is a fork - but the codebase is ... well a MEGA file:

Yes, I know. It grew over time and then I got used to it. I only started using Python with this project, bit by bit. In the next few days, I'm going to look into the "Sourcery" extension to "optimize" the code a bit. At the moment I don't have the knowledge to split the code into several files and what needs to be considered.

**Comment by jokob-sk** [comment]
For inspiration, this is how the code is split up in my fork: https://github.com/jokob-sk/Pi.Alert/tree/main/pialert - not sure how much of it can be re-used here.

**Comment by jeeftor** [comment]
Could probably drop the whole thing into ChatGPT and have it refactor for you

**Comment by leiweibau** [comment]
> Could probably drop the whole thing into ChatGPT and have it refactor for you

Sourcery is also AI-based. I prefer to do this selectively for the time being so that nothing blows up. I've already managed to save 60 lines 👍

I tried to move a few functions to another script, but that doesn't really work yet.

**Comment by leiweibau** [comment]
Thanks to the donations, I have ordered a cheap Asus router and will try to implement an importer. It remains to be seen whether I will succeed. And whether a single router model is sufficient to cover all Asus routers is the next question.

**Comment by leiweibau** [comment]
I have tested with only one router model (RT-AX53U). However, I also admit that I didn't do more intensive testing as the other tweaks and fixes were increasing. I will do some more testing in the next few days. Perhaps there will be one or two others who also own an Asus router.

**Comment by jeeftor** [comment]
I moved on to Opnsense ... :)

**Comment by leiweibau** [comment]
That's okay. I'm looking for general feedback, not yours in particular. After more than a year, I figured there was no longer an acute need 😉

**Comment by mdima** [comment]
Hello, sorry for the delay. 
I just activated it, the configuration is OK but I receive a "...Skipped. Could not connect to Asus Router" in the scan log. 

Where can I get some more information in the logs to dig a bit into this? 

Thanks,
Michele

**Comment by leiweibau** [comment]
The next update will have a button in the settings, within the "Log Viewer" box, with which you can activate the extended logs. I will also have added additional sections to the extended logging with the update. I hope that we can narrow down your problem with this.

**Comment by mdima** [comment]
> The next update will have a button in the settings, within the "Log Viewer" box, with which you can activate the extended logs. I will also have added additional sections to the extended logging with the update. I hope that we can narrow down your problem with this.

Ok, thank you very much.
